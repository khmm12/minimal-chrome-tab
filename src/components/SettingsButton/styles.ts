import { css } from 'styled-system/css'

export const button = css.raw({
  width: '2rem',
  height: '2rem',
  flex: '0 0 auto',
  color: { base: 'neutral.700', _dark: 'neutral.300', _hover: { base: 'black', _dark: 'white' } },
  appearance: 'none',
  padding: '0.25rem',
  background: 'transparent',
  border: 'none',
  cursor: { base: 'pointer', _disabled: 'default' },
  transition: '0.1s color ease-out',
  fontSize: '1.5rem',
})

export const svg = css.raw({
  transform: { base: 'rotate(0deg)', _groupHover: 'rotate(22.5deg)' },
  transition: '0.1s transform ease-out',
})
