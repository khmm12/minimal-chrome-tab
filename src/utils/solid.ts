export type SignalValue<T> = () => T

export function unwrap<T>(mayBeSingnal: SignalValue<T> | Exclude<T, Function>): T {
  // @ts-expect-error
  return typeof mayBeSingnal === 'function' ? mayBeSingnal() : mayBeSingnal
}
