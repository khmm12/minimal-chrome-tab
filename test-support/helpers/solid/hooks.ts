import { createRoot } from 'solid-js'

type Disposal = () => void

const disposals = new Set<Disposal>()

afterEach(() => cleanup())

export function renderHook<T>(fn: () => T): T {
  return withRoot(fn)[0]
}

export function withRoot<T>(fn: () => T): [value: T, dispose: () => void] {
  let disposal: () => void

  const value = createRoot((dispose) => {
    disposal = dispose
    disposals.add(dispose)
    return fn()
  })

  const dispose = (): void => {
    disposal()
    disposals.delete(disposal)
  }

  return [value, dispose]
}

export function cleanup(): void {
  for (const dispose of disposals) dispose()
  disposals.clear()
}
