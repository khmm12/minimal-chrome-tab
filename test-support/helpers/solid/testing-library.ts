import type { JSX } from 'solid-js'
import { hydrate as solidHydrate, render as solidRender } from 'solid-js/web'
import { getQueriesForElement, prettyDOM, queries as allQueries } from '@testing-library/dom'
import type { Queries, BoundFunction, prettyFormat } from '@testing-library/dom'

interface Ref {
  container: HTMLElement
  unmount: () => void
  dispose: () => void
}

type Ui = () => JSX.Element

interface Options<Q extends Queries = typeof allQueries> {
  container?: HTMLElement
  baseElement?: HTMLElement
  queries?: Q
  hydrate?: boolean
}

type DebugFn = (
  baseElement?: HTMLElement | HTMLElement[],
  maxLength?: number,
  options?: prettyFormat.OptionsReceived
) => void

type Result<Q extends Queries = typeof allQueries> = {
  container: HTMLElement
  baseElement: HTMLElement
  debug: DebugFn
  unmount: () => void
} & { [P in keyof Q]: BoundFunction<Q[P]> }

afterEach(() => cleanup())

const mountedContainers = new Set<Ref>()

function render<Q extends Queries = typeof allQueries>(ui: Ui, options: Options<Q>): Result<Q>
function render(ui: Ui, options?: Omit<Options, 'queries'>): Result

function render(ui: Ui, options: Options = {}): unknown {
  let { container, baseElement = container ?? document.body, queries, hydrate = false } = options

  container ??= baseElement.appendChild(document.createElement('div'))

  const unmount = (): void => {
    if (container?.parentElement === document.body) container.remove()
  }

  const dispose = hydrate ? solidHydrate(ui, container) : solidRender(ui, container)

  mountedContainers.add({ container, unmount, dispose })

  const debug: DebugFn = (el = baseElement, maxLength, options) =>
    Array.isArray(el)
      ? el.forEach((e) => console.log(prettyDOM(e, maxLength, options)))
      : console.log(prettyDOM(el, maxLength, options))

  return {
    container,
    baseElement,
    debug,
    unmount: dispose,
    ...getQueriesForElement(baseElement, queries),
  }
}

function cleanup(): void {
  mountedContainers.forEach(cleanupAtContainer)
  mountedContainers.clear()
}

function cleanupAtContainer(ref: Ref): void {
  ref.dispose()
  ref.unmount()
  mountedContainers.delete(ref)
}

export { render, cleanup }
