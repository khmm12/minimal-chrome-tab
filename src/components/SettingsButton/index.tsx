import { JSX, onCleanup } from 'solid-js'
import { SettingsIcon } from '@/components/Icon'
import * as css from './styles'

interface SettingsButtonProps {
  onClick?: () => void
}

export default function SettingsButton(props: SettingsButtonProps): JSX.Element {
  let $svg: SVGSVGElement | undefined
  let animation: Animation | undefined

  onCleanup(() => {
    animation?.cancel()
  })

  const handleClick = (): void => {
    if ($svg != null) {
      animation ??= animate($svg)
      animation?.play()
    }
    props.onClick?.()
  }

  return (
    <button class={css.button} type="button" title="Open settings" onClick={handleClick}>
      <SettingsIcon ref={$svg} aria-hidden class={css.svg} />
    </button>
  )
}

function animate($el: Element): Animation | undefined {
  if (typeof $el.animate === 'function')
    return $el.animate([{ transform: 'rotate(180deg)' }], {
      duration: 300,
      easing: 'ease-out',
    })
}
