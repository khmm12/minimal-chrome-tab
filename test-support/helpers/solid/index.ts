import { render, cleanup as cleanupElements } from './testing-library'
import { renderHook, withRoot, cleanup as cleanupHooks } from './hooks'

export { renderHook, withRoot, render }

export function cleanup(): void {
  cleanupHooks()
  cleanupElements()
}

export * from '@testing-library/dom'
