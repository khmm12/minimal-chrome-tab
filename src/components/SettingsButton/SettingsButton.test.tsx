import { createResource, createSignal, Show, Suspense } from 'solid-js'
import { fireEvent, render, renderHook, waitFor } from '@test/helpers/solid'
import SettingsButton from '.'

describe('SettingsButton', () => {
  it('renders a button', async () => {
    const handleClick = vi.fn()
    const r = render(() => <SettingsButton onClick={handleClick} />)

    const button = r.getByRole('button')

    expect(button).toBeInTheDocument()
    expect(button).toMatchSnapshot()
    expect(button).toHaveAttribute('aria-disabled', 'false')
    expect(button).toHaveAccessibleName('Open settings')

    fireEvent.click(button)

    await waitFor(() => {
      expect(handleClick).toBeCalled()
    })
  })

  it('handles deferred resources', async () => {
    const [isShown, setIsShown] = renderHook(() => createSignal(false)).result
    const [p, resolve] = deferred<null>()
    const [resource] = renderHook(() => createResource(async () => await p)).result

    const r = render(() => (
      <>
        <SettingsButton onClick={() => setIsShown(true)} />
        <Show when={isShown()}>
          <Suspense>
            <div>{resource()}</div>
          </Suspense>
        </Show>
      </>
    ))

    const button = r.getByRole('button')

    fireEvent.click(button)

    expect(button).toHaveAttribute('aria-disabled', 'true')
    expect(button).toHaveAccessibleName('Opening settings')

    resolve(null)

    await waitFor(() => {
      expect(button).toHaveAttribute('aria-disabled', 'false')
      expect(button).toHaveAccessibleName('Open settings')
    })
  })
})

function deferred<T>(): [p: Promise<T>, resolve: (val: T) => void] {
  let r: (val: T) => void

  const p = new Promise<T>((resolve) => {
    r = resolve
  })

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return [p, r!]
}
