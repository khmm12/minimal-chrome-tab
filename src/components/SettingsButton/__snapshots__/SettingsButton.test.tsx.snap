// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsButton > renders a button 1`] = `
<button
  aria-disabled="false"
  class="w_2rem h_2rem flex_0_0_auto c_neutral.700 dark:c_neutral.300 hover:c_black hover:dark:c_white ap_none p_0.25rem bg_transparent bd_none cursor_pointer disabled:cursor_default trs_0.1s_color_ease-out fs_1.5rem group"
  title="Open settings"
  type="button"
>
  <svg
    aria-hidden="true"
    class="w_1em h_1em trf_rotate(0deg) groupHover:trf_rotate(22.5deg) trs_0.1s_transform_ease-out"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="none"
      fill-rule="evenodd"
      stroke="currentColor"
      stroke-width="1.161"
    >
      <path
        d="M15.45.658c.743.226 1.456.523 2.13.883l-.262 5.141 5.141-.262c.36.674.657 1.386.883 2.13L19.521 12l3.821 3.45a11.775 11.775 0 0 1-.884 2.13l-5.14-.262.262 5.14c-.674.36-1.386.658-2.13.884L12 19.52l-3.45 3.822a11.775 11.775 0 0 1-2.13-.883l.262-5.141-5.14.263a11.775 11.775 0 0 1-.884-2.131L4.48 12 .658 8.55c.226-.744.523-1.456.883-2.13l5.141.262-.262-5.14C7.094 1.18 7.807.883 8.551.657l3.45 3.822Z"
      />
      <circle
        cx="11.963"
        cy="11.963"
        r="4.065"
      />
    </g>
  </svg>
</button>
`;
